src/util.rs

```diff
@@ -472,7 +492,58 @@ pub(crate) mod ptr {
         }
     }
 
-    impl<'a, T: 'a + ?Sized> Debug for Ptr<'a, T> {
+    impl<'a, T: 'a + ?Sized, const ASSUME_ALIGNED: bool> Debug for Ptr<'a, T, ASSUME_ALIGNED> {
+        #[inline]
+        fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+            self.ptr.fmt(f)
+        }
+    }
+
+    /// A reference to a validly-initialized `T` of lifetime `'a` that might
+    /// *not* be validly aligned for `T`.
+    #[doc(hidden)]
+    #[allow(unreachable_pub)] // false positive?
+    pub struct MaybeAligned<'a, T: ?Sized> {
+        /// INVARIANTS:
+        /// 1. `ptr` is validly-initialized for `T`
+        ptr: Ptr<'a, T, false>,
+    }
+
+    impl<'a, T> MaybeAligned<'a, T>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Read the value from `MaybeAligned`.
+        ///
+        /// This is only available if `T` is [`Copy`].
+        #[inline]
+        pub fn read_unaligned(self) -> T
+        where
+            T: Copy,
+        {
+            let raw = self.ptr.ptr.as_ptr();
+            // SAFETY: By invariant on `MaybeAligned`, `raw` contains
+            // validly-initialized data for `T`. The value is safe to read and
+            // return, because `T` is copy.
+            unsafe { core::ptr::read_unaligned(raw) }
+        }
+
+        /// View the value as an aligned reference.
+        ///
+        /// This is only available if `T` is [`Unaligned`].
+        #[inline]
+        pub fn as_ref(self) -> &'a T
+        where
+            T: Unaligned,
+        {
+            // SAFETY: By invariant on `MaybeAligned`, `ptr` is
+            // validly-initialized for `T`. Since `T` is `Unaligned`, it is also
+            // validly-aligned for `T`.
```

Comment:
What about aliasing? `Ptr` doesn't guarantee that other code accesses the same memory.